File: TanksNative/Bullet.cpp
#include "Bullet.h"
#include "LevelConstructor.h"
#include "ResourceManager.h"
#include "Log.h"
#include "Block.h"
#include "GameObjectManager.h"


Bullet::Bullet(Direction direction)
	: direction{ direction }, _Speed{ &BulletSpeed::Normal }
{
	SetSize({ LevelConstructor::GetMinBlockSize(),
		LevelConstructor::GetMinBlockSize() });
	SetTextureCoords(ResourceManager::GetBulletCoords(direction));
}

Bullet::~Bullet()
{
	Log::WriteLine(L"Bullet is detroyed!");
}

void Bullet::Update(float elapsedTime)
{
	// перемещаемся
	float units = (_Speed->GetValue()) * elapsedTime;
	if (direction == Direction::Top)
		SetPosition({ GetPosition().x, GetPosition().y - units });
	else if (direction == Direction::Bottom)
		SetPosition({ GetPosition().x, GetPosition().y + units });
	else if (direction == Direction::Right)
		SetPosition({ GetPosition().x + units, GetPosition().y });
	else if (direction == Direction::Left)
		SetPosition({ GetPosition().x - units, GetPosition().y });
}

void Bullet::OnCollisionDetected(std::shared_ptr<GameObject>& a, std::shared_ptr<GameObject>& b)
{
	/*auto block = std::dynamic_pointer_cast<Block>(another);
	if (block) {
		Log::WriteLine(L"Bullet is crushed!");
		auto wall = block->GetWall();
		if(wall == Walls.)
	}*/

	if (b->GetDepth() == &Depth::Action) {
		Log::WriteLine(L"Bullet is crushed!");
		GameObjectManager::Destroy(a);
	}
}

File: Tanks/Game/Bullet.cs
using SharpDX;

namespace Tanks.Game
{
	internal class Bullet : GameObject
	{
		public Direction Direction { get; }
		public BulletSpeed Speed { get; set; } = BulletSpeed.Normal;
		public delegate void BulletHandler(Bullet bullet);
		public event BulletHandler Destroy;



		public Bullet(Direction direction)
		{
			Direction = direction;
			Size = new Vector2(LevelConstructor.Instance.MinBlockSize, LevelConstructor.Instance.MinBlockSize);
			TextureCoords = ResourceManager.Instance.GetBulletCoords(direction);
		}

		public override void Update(float elapsedTime)
		{
			// перемещаемся
			float units = (Speed?.Value ?? BulletSpeed.Normal.Value) * elapsedTime;
			if (Direction == Direction.Top)
				Position = new SharpDX.Vector2(Position.X, Position.Y - units);
			else if (Direction == Direction.Bottom)
				Position = new SharpDX.Vector2(Position.X, Position.Y + units);
			else if (Direction == Direction.Right)
				Position = new SharpDX.Vector2(Position.X + units, Position.Y);
			else if (Direction == Direction.Left)
				Position = new SharpDX.Vector2(Position.X - units, Position.Y);
		}
	}

	class BulletSpeed
	{
		public float Value { get; set; }
		public static readonly BulletSpeed Normal = new BulletSpeed(100.0f);
		public static readonly BulletSpeed Fast = new BulletSpeed(200.0f);

		BulletSpeed(float v)
		{
			Value = v;
		}
	}
}